pkg/identity/cache/allocator.go

```diff
@@ -399,15 +434,158 @@ func (m *CachingIdentityAllocator) UnwithholdLocalIdentities(nids []identity.Num
 	m.localNodeIdentities.unwithhold(nids)
 }
 
+// checkpoint writes the state of the local allocators to disk. This is used for restoration,
+// to ensure that numeric identities are, as much as possible, stable across agent restarts.
+//
+// Do not call this directly, rather, use m.checkpointTrigger.Trigger()
+func (m *CachingIdentityAllocator) checkpoint(reasons []string) {
+	if m.checkpointPath == "" {
+		return // this is a unit test
+	}
+	log := log.WithField(logfields.Path, m.checkpointPath)
+
+	ids := make([]*identity.Identity, 0, m.localIdentities.size()+m.localNodeIdentities.size())
+	ids = m.localIdentities.checkpoint(ids)
+	ids = m.localNodeIdentities.checkpoint(ids)
+
+	// use renameio to prevent partial writes
+	out, err := renameio.NewPendingFile(m.checkpointPath, renameio.WithExistingPermissions(), renameio.WithPermissions(0o600))
+	if err != nil {
+		log.WithError(err).Error("failed to prepare checkpoint file")
+		return
+	}
+	defer out.Cleanup()
+
+	jw := jsoniter.ConfigFastest.NewEncoder(out)
+	if err := jw.Encode(ids); err != nil {
+		log.WithError(err).Error("failed to marshal identity checkpoint state")
+		return
+	}
+	if err := out.CloseAtomicallyReplace(); err != nil {
+		log.WithError(err).Error("failed to write identity checkpoint file")
+		return
+	}
+	log.Debug("Wrote local identity allocator checkpoint")
+}
+
+// RestoreLocalIdentities reads in the checkpointed local allocator state
+// from disk and allocates a reference to every previously existing identity.
+//
+// Once all identity-allocating objects are synchronized (e.g. network policies,
+// remote nodes), call ReleaseRestoredIdentities to release the held references.
+func (m *CachingIdentityAllocator) RestoreLocalIdentities() (map[identity.NumericIdentity]*identity.Identity, error) {
+	if m.checkpointPath == "" {
+		return nil, nil // unit test
+	}
+	log := log.WithField(logfields.Path, m.checkpointPath)
+
+	fp, err := os.Open(m.checkpointPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			log.Info("No identity checkpoint file found, skipping restoration")
+			return nil, nil
+		}
+		return nil, fmt.Errorf("failed to open identity checkpoint file %s: %w", m.checkpointPath, err)
+	}
+	defer fp.Close()
+
+	jr := jsoniter.ConfigFastest.NewDecoder(fp)
+	var ids []*identity.Identity
+	if err := jr.Decode(&ids); err != nil {
+		return nil, fmt.Errorf("failed to parse identity checkpoint file %s: %w", m.checkpointPath, err)
+	}
+
+	if len(ids) == 0 {
+		return nil, nil
+	}
+
+	log.WithField(logfields.Count, len(ids)).Info("Restoring checkpointed local identities")
+
+	m.restoredIdentities = make(map[identity.NumericIdentity]*identity.Identity, len(ids))
+
+	// create an immediately-cancelled Context for use during ID restoration.
+	// We will use this to ensure we never do remote identity allocation.
+	closedCtx, cfunc := context.WithCancel(context.Background())
+	cfunc()
```

Comment:
This doesn't spark joy. I understand that changing the allocator interface would cause lots of churn. 

Since caller and callee are in the same struct, is there no way we move the "allocate local identity" part out of `AllocateIdentity` and have it be a private helper called both from here and `AllocateIdentity`?

Reply:
You're right about this for sure. Let me think about it.

To be honest, we ultimately want to change the interface anyways. `ipcache.resolveIdentity()` should never allocate non-local identities anyways, so we need the option. Let me add another method.

Reply:
I added `AllocateLocalIdentity()`, which we will need regardless.